Fix missing allocations in `memray stats` histogram

xlanor opened the pull request:
Signed-off-by: Jingkai Tan <tjingkai@bloomberg.net>

Fixes #84.

**Describe your changes**
Cap the known valid range to ensure that anything greater than the second to last bucket will appear in the last bucket.

**Testing performed**
Ensured that output is as expected (missing mmap allocation shows up)
```
🕙12:30:27 ❯ memray stats output.bin
📏 Total allocations:
	9

📦 Total memory allocated:
	9.313GB

📊 Histogram of allocation size:
	min: 536.000B
	----------------------------------------
	< 2.792KB  : 4 ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇
	< 14.895KB : 1 ▇▇▇▇▇▇▇
	< 79.453KB : 0
	< 423.837KB: 0
	< 2.208MB  : 0
	< 11.778MB : 0
	< 62.827MB : 0
	< 335.145MB: 0
	< 1.746GB  : 0
	<=9.313GB  : 1 ▇▇▇▇▇▇▇
	----------------------------------------
	max: 9.313GB

📂 Allocator type distribution:
	 MALLOC: 5
	 MMAP: 1

🥇 Top 5 largest allocating locations (by size):
	- <module>:../test.py:3 -> 9.313GB
	- _call_with_frames_removed:<frozen importlib._bootstrap>:219 -> 4.381KB
	- _run_tracker:/home/jingkai/Documents/github/memray/memray/src/memray/commands/run.py:60 -> 592.000B
	- _run_code:/usr/lib/python3.8/runpy.py:80 -> 576.000B
	- run_path:/usr/lib/python3.8/runpy.py:265 -> 536.000B

🥇 Top 5 largest allocating locations (by number of allocations):
	- _call_with_frames_removed:<frozen importlib._bootstrap>:219 -> 4
	- <module>:../test.py:3 -> 1
	- run_path:/usr/lib/python3.8/runpy.py:265 -> 1
	- _run_module_code:/usr/lib/python3.8/runpy.py:97 -> 1
	- _run_tracker:/home/jingkai/Documents/github/memray/memray/src/memray/commands/run.py:60 -> 1
```

godlygeek commented on src/memray/reporters/stats.py:
```diff
@@ -90,12 +90,13 @@ def get_histogram_databins(data: List[int], bins: int) -> List[Tuple[int, int]]:
     # Determine the upper bound in bytes for each bin
-    steps = [int(math.exp(low + step * (i + 1))) for i in range(bins)]
+    steps = [int(math.exp(low + step * (i + 1))) for i in range(bins+1)]
     dist = Counter((x - low) // step for x in it)
```
I think a better solution would be clamping `(x - low) // step` to the known valid range, so that anything >= the 2nd to last bucket automatically goes in the last bucket. Does:
```py
    dist = Counter(min((x - low) // step, bins - 1) for x in it)
```
solve the problem? That should make it so that no element in the counter ever has a value greater than `bins - 1`.

xlanor replied:
Definetly, you're right. Let me make the amendments

xlanor commented:
Could I get an approval to rerun CI on this please?

godlygeek reviewed (approved):
I've made a few minor tweaks - rebased this, ran Black, added a changelog entry, and squashed it down to one commit. Looks good to me!

Thanks @xlanor, especially for the new test to try to make sure we don't regress on this again in the future.

xlanor commented:
> I've made a few minor tweaks - rebased this, ran Black, added a changelog entry, and squashed it down to one commit. Looks good to me!
> 
> 
> 
> Thanks @xlanor, especially for the new test to try to make sure we don't regress on this again in the future.

Thank you Matt and Pablo for your guidance on this PR. I hope to contribute more to this project in the future during my free time!

pablogsal commented:
> > I've made a few minor tweaks - rebased this, ran Black, added a changelog entry, and squashed it down to one commit. Looks good to me!
> > Thanks @xlanor, especially for the new test to try to make sure we don't regress on this again in the future.
> 
> Thank you Matt and Pablo for your guidance on this PR. I hope to contribute more to this project in the future during my free time!

Thanks a lot for your great work and patience ❤️

xlanor commented:
> > > I've made a few minor tweaks - rebased this, ran Black, added a changelog entry, and squashed it down to one commit. Looks good to me!
> 
> > > Thanks @xlanor, especially for the new test to try to make sure we don't regress on this again in the future.
> 
> > 
> 
> > Thank you Matt and Pablo for your guidance on this PR. I hope to contribute more to this project in the future during my free time!
> 
> 
> 
> Thanks a lot for your great work and patience ❤️ 

No problem! Please feel free to let me know if you're in the London office and want to grab a drink.
